File: nf-dream/scripts/consolidate.py
#!/usr/bin/env python3
"""nf-dream consolidate — move aged session-end states + Stale bucket into archive/.

Reads classification from classify.py (run automatically). Builds move plan + executes
with --apply. Updates archive/README.md breadcrumb and MEMORY.md index.

Usage:
  consolidate.py <scope> [--apply] [--json] [--project <slug>] [--age-days 14]
"""
from __future__ import annotations

import argparse
import json
import re
import subprocess
import sys
import time
from datetime import datetime
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from _lib import iter_memory_files, parse_frontmatter, extract_frontmatter_dict, is_session_end  # noqa: E402


def run_classify(scope: Path, project: str | None) -> dict:
    cmd = [sys.executable, str(Path(__file__).parent / "classify.py"), str(scope), "--json"]
    if project:
        cmd += ["--project", project]
    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        raise RuntimeError(f"classify.py failed: {result.stderr}")
    return json.loads(result.stdout)


def derive_archive_month(file_path: Path) -> str:
    """Extract YYYY-MM from filename like project_session_end_state_2026_05_14.md, else use mtime."""
    name = file_path.name
    m = re.search(r"(\d{4})[_-](\d{2})", name)
    if m:
        return f"{m.group(1)}-{m.group(2)}"
    mtime = file_path.stat().st_mtime
    return datetime.fromtimestamp(mtime).strftime("%Y-%m")


def build_plan(scope: Path, classify_data: dict, age_days: int) -> list[dict]:
    age_cutoff = time.time() - age_days * 24 * 3600
    plan: list[dict] = []
    for bucket in classify_data["buckets"].values():
        for entry in bucket:
            path = scope / entry["file"]
            if not path.exists():
                continue
            archive_candidate = False
            reason = ""
            if entry["bucket"] == "Stale":
                archive_candidate = True
                reason = f"Stale ({entry['reason']})"
            elif is_session_end(entry["file"]) and entry["mtime"] < age_cutoff:
                archive_candidate = True
                reason = f"session-end older than {age_days}d"
            if archive_candidate:
                ym = derive_archive_month(path)
                plan.append({
                    "src": entry["file"],
                    "target": f"archive/{ym}/{entry['file']}",
                    "bucket": entry["bucket"],
                    "reason": reason,
                })
    return plan


def update_archive_readme(scope: Path, moves: list[dict]):
    readme = scope / "archive" / "README.md"
    readme.parent.mkdir(exist_ok=True)
    today = datetime.now().strftime("%Y-%m-%d")
    existing = readme.read_text() if readme.exists() else "# Archive\n\nMoved memory files by date.\n"
    lines = [existing.rstrip(), ""]
    for m in moves:
        lines.append(f"- {today}: `{m['src']}` → `{m['target']}` ({m['reason']})")
    lines.append("")
    readme.write_text("\n".join(lines))


def update_memory_md(scope: Path, archived_names: set[str]):
    memory = scope / "MEMORY.md"
    if not memory.exists():
        return
    text = memory.read_text()
    new_lines: list[str] = []
    removed = 0
    for line in text.split("\n"):
        m = re.search(r"\[[^\]]+\]\(([^)]+\.md)\)", line)
        if m and m.group(1) in archived_names:
            removed += 1
            continue
        new_lines.append(line)
    # Ensure archive index entry exists
    new_text = "\n".join(new_lines)
    if "- [Archive](archive/README.md)" not in new_text:
        # Append at end before EOF
        new_text = new_text.rstrip() + "\n\n- [Archive](archive/README.md) — consolidated session-end states\n"
    memory.write_text(new_text)
    return removed


def apply_plan(scope: Path, plan: list[dict]) -> dict:
    moves_done: list[dict] = []
    for entry in plan:
        src = scope / entry["src"]
        target = scope / entry["target"]
        target.parent.mkdir(parents=True, exist_ok=True)
        src.rename(target)
        moves_done.append(entry)
    if moves_done:
        update_archive_readme(scope, moves_done)
        update_memory_md(scope, {m["src"] for m in moves_done})
    return {"ok": True, "moved": len(moves_done)}


def main() -> int:
    parser = argparse.ArgumentParser(description="nf-dream consolidate")
    parser.add_argument("scope")
    parser.add_argument("--apply", action="store_true")
    parser.add_argument("--json", action="store_true")
    parser.add_argument("--project")
    parser.add_argument("--age-days", type=int, default=14)
    args = parser.parse_args()

    scope = Path(args.scope).expanduser().resolve()
    if not scope.is_dir():
        print(f"ERROR: {scope} is not a directory", file=sys.stderr)
        return 1

    classify_data = run_classify(scope, args.project)
    plan = build_plan(scope, classify_data, args.age_days)
    summary = {
        "scope": str(scope),
        "applied": args.apply,
        "candidates": len(plan),
        "by_target": {},
    }
    for m in plan:
        ym = m["target"].split("/")[1]
        summary["by_target"][ym] = summary["by_target"].get(ym, 0) + 1

    if args.apply and plan:
        apply_plan(scope, plan)

    if args.json:
        print(json.dumps({"summary": summary, "plan": plan}, indent=2))
        return 0

    print(f"nf-dream consolidate — {scope}{' (APPLIED)' if args.apply else ' (dry-run)'}")
    print(f"  candidates: {len(plan)}")
    if not plan:
        print("  nothing to consolidate")
        return 0
    for ym, n in sorted(summary["by_target"].items()):
        print(f"  → archive/{ym}/: {n} files")
    print()
    print("Plan:")
    for m in plan[:30]:
        print(f"  {m['src']:50s} → {m['target']}  ({m['reason']})")
    if len(plan) > 30:
        print(f"  ... and {len(plan) - 30} more")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: nf-dream/scripts/consolidate.ts
#!/usr/bin/env bun
/**
 * nf-dream consolidate — move aged session-end states + Stale bucket into
 * archive/YYYY-MM/. Updates archive/README.md breadcrumb + MEMORY.md index.
 *
 * Usage:
 *   bun consolidate.ts <scope> [--apply] [--json] [--project <slug>]
 *                              [--include-generic] [--age-days 14]
 */
import {
  readFileSync, writeFileSync, existsSync, statSync, mkdirSync, renameSync,
} from "node:fs";
import { resolve, dirname } from "node:path";
import { isSessionEnd } from "./_lib";
import { classifyScope } from "./classify";

function deriveArchiveMonth(path: string): string {
  const name = path.split("/").pop() ?? "";
  const m = name.match(/(\d{4})[_-](\d{2})/);
  if (m) return `${m[1]}-${m[2]}`;
  const mtime = statSync(path).mtime;
  return `${mtime.getFullYear()}-${String(mtime.getMonth() + 1).padStart(2, "0")}`;
}

interface PlanItem {
  src: string;
  target: string;
  bucket: string;
  reason: string;
}

interface BuildOpts {
  scope: string;
  project: string | null;
  includeGeneric: boolean;
  ageDays: number;
}

function buildPlan(opts: BuildOpts): PlanItem[] {
  const classifyData = classifyScope(opts.scope, opts.project, opts.includeGeneric);
  const ageCutoff = Date.now() / 1000 - opts.ageDays * 24 * 3600;
  const plan: PlanItem[] = [];
  for (const entries of Object.values(classifyData.buckets)) {
    for (const entry of entries) {
      const path = resolve(opts.scope, entry.file);
      if (!existsSync(path)) continue;
      let archiveCandidate = false;
      let reason = "";
      if (entry.bucket === "Stale") {
        archiveCandidate = true;
        reason = `Stale (${entry.reason})`;
      } else if (isSessionEnd(entry.file) && entry.mtime < ageCutoff) {
        archiveCandidate = true;
        reason = `session-end older than ${opts.ageDays}d`;
      }
      if (archiveCandidate) {
        const ym = deriveArchiveMonth(path);
        plan.push({
          src: entry.file,
          target: `archive/${ym}/${entry.file}`,
          bucket: entry.bucket,
          reason,
        });
      }
    }
  }
  return plan;
}

function updateArchiveReadme(scope: string, moves: PlanItem[]): void {
  const readmePath = resolve(scope, "archive", "README.md");
  mkdirSync(dirname(readmePath), { recursive: true });
  const today = new Date().toISOString().slice(0, 10);
  const existing = existsSync(readmePath)
    ? readFileSync(readmePath, "utf8")
    : "# Archive\n\nMoved memory files by date.\n";
  const lines = [existing.trimEnd(), ""];
  for (const m of moves) {
    lines.push(`- ${today}: \`${m.src}\` → \`${m.target}\` (${m.reason})`);
  }
  lines.push("");
  writeFileSync(readmePath, lines.join("\n"));
}

function updateMemoryMd(scope: string, archivedNames: Set<string>): number {
  const memoryPath = resolve(scope, "MEMORY.md");
  if (!existsSync(memoryPath)) return 0;
  const text = readFileSync(memoryPath, "utf8");
  const newLines: string[] = [];
  let removed = 0;
  for (const line of text.split("\n")) {
    const m = line.match(/\[[^\]]+\]\(([^)]+\.md)\)/);
    if (m && archivedNames.has(m[1])) { removed++; continue; }
    newLines.push(line);
  }
  let newText = newLines.join("\n");
  if (!newText.includes("- [Archive](archive/README.md)")) {
    newText = newText.trimEnd() + "\n\n- [Archive](archive/README.md) — consolidated session-end states\n";
  }
  writeFileSync(memoryPath, newText);
  return removed;
}

function applyPlan(scope: string, plan: PlanItem[]): { ok: true; moved: number } {
  const movesDone: PlanItem[] = [];
  for (const entry of plan) {
    const src = resolve(scope, entry.src);
    const target = resolve(scope, entry.target);
    mkdirSync(dirname(target), { recursive: true });
    renameSync(src, target);
    movesDone.push(entry);
  }
  if (movesDone.length > 0) {
    updateArchiveReadme(scope, movesDone);
    updateMemoryMd(scope, new Set(movesDone.map((m) => m.src)));
  }
  return { ok: true, moved: movesDone.length };
}

interface Args {
  scope: string;
  apply: boolean;
  json: boolean;
  project: string | null;
  ageDays: number;
  includeGeneric: boolean;
  help: boolean;
}

function parseArgs(argv: string[]): Args {
  const a: Args = {
    scope: "", apply: false, json: false, project: null,
    ageDays: 14, includeGeneric: false, help: false,
  };
  const rest: string[] = [];
  for (let i = 0; i < argv.length; i++) {
    const t = argv[i];
    if (t === "--help" || t === "-h") a.help = true;
    else if (t === "--apply") a.apply = true;
    else if (t === "--json") a.json = true;
    else if (t === "--include-generic") a.includeGeneric = true;
    else if (t === "--project") a.project = argv[++i] ?? null;
    else if (t === "--age-days") a.ageDays = parseInt(argv[++i] ?? "14", 10);
    else if (t.startsWith("--")) throw new Error(`unknown flag: ${t}`);
    else rest.push(t);
  }
  if (rest.length === 0 && !a.help) throw new Error("missing <scope>");
  a.scope = rest[0] ?? "";
  return a;
}

function printHelp(): void {
  console.log(`bun consolidate.ts <scope> [--apply] [--json] [--project <slug>]
                          [--include-generic] [--age-days 14]

Move aged session-end + Stale bucket into archive/YYYY-MM/.`);
}

async function main(): Promise<number> {
  let args: Args;
  try {
    args = parseArgs(process.argv.slice(2));
  } catch (err) {
    console.error(`ERROR: ${(err as Error).message}`);
    printHelp();
    return 1;
  }
  if (args.help) { printHelp(); return 0; }
  const scope = resolve(args.scope.replace(/^~/, process.env.HOME ?? ""));
  try {
    if (!statSync(scope).isDirectory()) throw new Error();
  } catch {
    console.error(`ERROR: ${scope} is not a directory`);
    return 1;
  }

  const plan = buildPlan({
    scope, project: args.project, includeGeneric: args.includeGeneric, ageDays: args.ageDays,
  });
  const byTarget: Record<string, number> = {};
  for (const m of plan) {
    const ym = m.target.split("/")[1];
    byTarget[ym] = (byTarget[ym] ?? 0) + 1;
  }
  const summary = {
    scope, applied: args.apply, candidates: plan.length, by_target: byTarget,
  };

  if (args.apply && plan.length > 0) applyPlan(scope, plan);

  if (args.json) {
    console.log(JSON.stringify({ summary, plan }, null, 2));
    return 0;
  }

  console.log(`nf-dream consolidate — ${scope}${args.apply ? " (APPLIED)" : " (dry-run)"}`);
  console.log(`  candidates: ${plan.length}`);
  if (plan.length === 0) { console.log("  nothing to consolidate"); return 0; }
  for (const [ym, n] of Object.entries(byTarget).sort()) {
    console.log(`  → archive/${ym}/: ${n} files`);
  }
  console.log();
  console.log("Plan:");
  for (const m of plan.slice(0, 30)) {
    console.log(`  ${m.src.padEnd(50)} → ${m.target}  (${m.reason})`);
  }
  if (plan.length > 30) console.log(`  ... and ${plan.length - 30} more`);
  return 0;
}

if (import.meta.main) {
  main().then((code) => process.exit(code));
}
